### ISSUE TITLE
Rounding issue on long values.

### ISSUE THREAD

**Issue by javadevmtl** [elastic/elasticsearch#15292]
Hi using ES 2.1.0

We have noticed a rounding issue with "big" long values.

I have provided a text file with the bulk operations executed and the search result of the index. You will notice that the value inserted does not match the initial bulk operation. It seems to be rounding/doing something funny.

We used default settings on the index. The mapping is detected as long.

[elastic-2.1.0-rounding-issue.txt](https://github.com/elastic/elasticsearch/files/54618/elastic-2.1.0-rounding-issue.txt)

**Comment by rjernst** [comment]
The json spec has precision issues with large longs, because it represents all numbers as doubles. Set up your mappings (don't use dynamic mappings) as long, and pass the value in a json string. The `coerce` setting (which defaults to true), will use Long.parseLong on the string value.

**Comment by javadevmtl** [comment]
Ok but the source will still shows as string when I return the search result?

Just I thought. Unless I missed it, the docs don't explain this phenomenon.

**Comment by clintongormley** [comment]
To be clear, this isn't a problem with the JSON spec per se, which says nothing about precision.  It is javascript which represents all numbers as doubles (and so a number of json encoders will do the same thing).  

Take your example:

    PUT t
    {
      "mappings": {
        "mytype": {
          "properties": {
            "my_huge": {
              "type": "long"
            }
          }
        }
      }
    }
    
    POST t/mytype/_bulk
    {"index": {"_id": "10000000000001487" }}
    {"my_huge":10000000000001487}
    {"index": {"_id": "10000000000001489" }}
    {"my_huge":10000000000001489}
    
    GET t/_search?pretty
    {
      "aggs": {
        "myhuge": {
          "terms": {
            "field": "my_huge"
          }
        }
      }
    }

If you run the above in Sense, it will output:

    "hits": {
      "total": 2,
      "max_score": 1,
      "hits": [
        {
          "_index": "t",
          "_type": "mytype",
          "_id": "10000000000001487",
          "_score": 1,
          "_source": {
            "my_huge": 10000000000001488
          }
        },
        {
          "_index": "t",
          "_type": "mytype",
          "_id": "10000000000001489",
          "_score": 1,
          "_source": {
            "my_huge": 10000000000001488
          }
        }
      ]
    },
    "aggregations": {
      "myhuge": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
          {
            "key": 10000000000001488,
            "doc_count": 1
          },
          {
            "key": 10000000000001488,
            "doc_count": 1
          }
        ]
      }
 
The number have all been rounded to 10000000000001488.

However, repeat the same search request in the console with curl and you get:

    "hits" : {
      "total" : 2,
      "max_score" : 1.0,
      "hits" : [ {
        "_index" : "t",
        "_type" : "mytype",
        "_id" : "10000000000001487",
        "_score" : 1.0,
        "_source":{"my_huge":10000000000001487}
      }, {
        "_index" : "t",
        "_type" : "mytype",
        "_id" : "10000000000001489",
        "_score" : 1.0,
        "_source":{"my_huge":10000000000001489}
      } ]
    },
    "aggregations" : {
      "myhuge" : {
        "doc_count_error_upper_bound" : 0,
        "sum_other_doc_count" : 0,
        "buckets" : [ {
          "key" : 10000000000001487,
          "doc_count" : 1
        }, {
          "key" : 10000000000001489,
          "doc_count" : 1
        } ]
      }
    }

In other words, Elasticsearch is doing the right thing, and the Sense example is rounded because Javascript parsed the response JSON.

**Comment by huuhau95** [comment]
a
